wallet/types/src/partially_signed_transaction.rs

```diff
@@ -13,50 +13,100 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use super::{
-    signature::{inputsig::InputWitness, Signable, Transactable},
-    Destination, Transaction, TxOutput,
+use common::{
+    chain::{
+        signature::{inputsig::InputWitness, Signable, Transactable},
+        Destination, SignedTransaction, Transaction, TransactionCreationError, TxInput, TxOutput,
+    },
+    primitives::Amount,
 };
-use crate::chain::{SignedTransaction, TransactionCreationError, TxInput};
 use serialization::{Decode, Encode};
+use thiserror::Error;
+use tx_verifier::input_check::signature_only_check::SignatureOnlyVerifiable;
 use utils::ensure;
 
+#[derive(Error, Debug, Clone, PartialEq, Eq)]
+pub enum PartiallySignedTransactionCreationError {
+    #[error("Failed to convert partially signed tx to signed")]
+    FailedToConvertPartiallySignedTx(PartiallySignedTransaction),
+    #[error("The number of output additional infos does not match the number of outputs")]
+    InvalidOutputAdditionalInfoCount,
+    #[error("Failed to create transaction: {0}")]
+    TxCreationError(#[from] TransactionCreationError),
+}
+
+/// Additional info for UTXOs
+#[derive(Debug, Eq, PartialEq, Clone, Encode, Decode)]
+pub enum UtxoAdditionalInfo {
+    TokenInfo { num_decimals: u8, ticker: Vec<u8> },
+    PoolInfo { staker_balance: Amount },
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Encode, Decode)]
+pub struct UtxoWithAdditionalInfo {
+    pub utxo: TxOutput,
+    pub additional_info: Option<UtxoAdditionalInfo>,
+}
+
+impl UtxoWithAdditionalInfo {
+    pub fn new(utxo: TxOutput, additional_info: Option<UtxoAdditionalInfo>) -> Self {
+        Self {
+            utxo,
+            additional_info,
+        }
+    }
+}
+
 #[derive(Debug, Eq, PartialEq, Clone, Encode, Decode)]
 pub struct PartiallySignedTransaction {
     tx: Transaction,
     witnesses: Vec<Option<InputWitness>>,
 
-    input_utxos: Vec<Option<TxOutput>>,
+    input_utxos: Vec<Option<UtxoWithAdditionalInfo>>,
```

Comment:
why not storing utxo separately from info and zipping if necessary? Right now the data is duplicated. Also it looks cleaner with all the data inside ptx flatten. If such new type is required for trezor I'd move it closer to it.

Reply:
I went with this to preserve the invariant that you can't have additional info if you don't have an UTXO as well. Also the none case of the optional for the additional info does not mean missing it means no additional info. Maybe I should go back to the enum instead of the option + enum?

Reply:
Now I get it. I'd leave a comment about that invariant and new field just to make it clear for the future.